src/adapter/src/optimize/dataflows.rs

```diff
@@ -131,23 +131,41 @@ pub struct DataflowBuilder<'a> {
 }
 
 /// The styles in which an expression can be prepared for use in a dataflow.
-#[derive(Clone, Copy, Debug)]
-pub enum ExprPrepStyle<'a> {
-    /// The expression is being prepared for installation as a maintained dataflow, e.g.,
-    /// index, materialized view, or subscribe.
-    Maintained,
-    /// The expression is being prepared to run once at the specified logical
-    /// time in the specified session.
-    OneShot {
-        logical_time: EvalTime,
-        session: &'a dyn SessionMetadata,
-        catalog_state: &'a CatalogState,
-    },
-    /// The expression is being prepared for evaluation in a CHECK expression of a webhook source.
-    WebhookValidation {
-        /// Time at which this expression is being evaluated.
-        now: DateTime<Utc>,
-    },
+pub trait ExprPrepStyle {
+    /// Prepare a relation expression according to this style.
+    fn prep_relation_expr(&self, expr: &mut OptimizedMirRelationExpr)
+    -> Result<(), OptimizerError>;
+
+    /// Prepare a scalar expression according to this style.
+    fn prep_scalar_expr(&self, expr: &mut MirScalarExpr) -> Result<(), OptimizerError>;
+}
+
+/// A no-op expression preparer.
+pub struct NoopExprPrepStyle;
+impl ExprPrepStyle for NoopExprPrepStyle {
+    fn prep_relation_expr(&self, _: &mut OptimizedMirRelationExpr) -> Result<(), OptimizerError> {
+        Ok(())
+    }
+    fn prep_scalar_expr(&self, _expr: &mut MirScalarExpr) -> Result<(), OptimizerError> {
+        Ok(())
+    }
+}
+
+/// Preparing an expression for maintained dataflow, e.g., index, materialized view, or subscribe.
+/// Produces errors for calls to an unmaterializable functions.
+pub struct ExprPrepStyleMaintained;
```

Comment:
Another nit: I'd put the impl blocks directly below their respective `struct` definitions, so it's easy to verify that the implementation matches the documented behavior.